test/distributed/test_c10d.py

```diff
@@ -3008,37 +3011,42 @@ def test_ddp_comm_hook_future_passing_cpu(self):
 
         # Test on CPU
         cpu_model = DistributedDataParallel(
-            TestDdpCommHook().cpu(),
-            process_group=process_group
+            TestDdpCommHook().cpu(), process_group=process_group
         )
 
         # Register DDP Communication Hook
         cpu_model._register_comm_hook(None, self._simple_hook)
 
-        self._run_and_verify_simple_hook(cpu_model)
-
-    def _test_ddp_comm_hook_future_passing_gpu(self, c10d_process_group):
-        """
-        This unit test verifies whether the Future object is passed properly.
-        The callback function creates a Future object and sets a value to it.
-        """
-        store = c10d.FileStore(self.file_name, self.world_size)
-        process_group = c10d_process_group(store, self.rank, self.world_size)
+        # check whether the grads are equal to what then callback returns.
+        # without the comm_hook, result would be 0.25 * torch.ones(2, 2).
+        self._run_and_verify_hook(cpu_model, 8, 2 * torch.ones(2, 2))
 
-        # Test on GPU
+    def _gpu_model_with_ddp_comm_hook(self, process_group, hook=None):
         device_id = gpus_for_rank(self.world_size)[self.rank][0]
         gpu_model = DistributedDataParallel(
             TestDdpCommHook().to(device_id),
             device_ids=[device_id],
-            process_group=process_group
+            process_group=process_group,
         )
 
-        # Register DDP Communication Hook
-        gpu_model._register_comm_hook(None, self._simple_hook)
+        # Register DDP Communication Hook if defined
+        if hook is not None:
+            gpu_model._register_comm_hook(None, hook)
+
+        return gpu_model
+
+    def _run_and_verify_hook(self, model, input, expected_grad):
+        # Run forward
+        output = model(input, self.rank)
 
-        self._run_and_verify_simple_hook(gpu_model)
+        # Run backward
+        output.mean().backward()
 
-    def _simple_hook(self, state: object, bucket: dist.GradBucket) -> torch.futures.Future:
+        [self.assertEqual(p.grad, expected_grad) for p in model.parameters()]
+
+    def _simple_hook(
+        self, state: object, bucket: dist._GradBucket
+    ) -> torch.futures.Future:
```

Comment:
Do we also have tests to cover the case where the hook directly returns the `FutureNCCL` acquired from `.get_future()`? Instead of chaining another `.then()`?

Reply:
Yes we do. Please see `test_ddp_comm_hook_allreduce_hook_nccl`:
https://github.com/pytorch/pytorch/blob/b49167474549b61f5fb554bfac6ef681e9b41654/test/distributed/test_c10d.py#L3105-L3106